apis/Google.Cloud.Firestore.Data/Google.Cloud.Firestore.Data.Tests/FirestoreDbTest.RunTransactionAsync.cs

```diff
@@ -0,0 +1,229 @@
+            Assert.Equal(3, result);
+
+            var expectedCommitRequests = new[]
+            {
+                CreateCommitRequest("transaction 1"),
+                CreateCommitRequest("transaction 2; retrying transaction 1"),
+                CreateCommitRequest("transaction 3; retrying transaction 2")
+            };
+            var expectedRollbackRequests = new[]
+            {
+                CreateRollbackRequest("transaction 1"),
+                CreateRollbackRequest("transaction 2; retrying transaction 1"),
+            };
+            Assert.Equal(expectedCommitRequests, client.CommitRequests);
+            Assert.Equal(expectedRollbackRequests, client.RollbackRequests);
+        }
+
+        [Theory]
+        [InlineData(StatusCode.DeadlineExceeded)]
+        [InlineData(StatusCode.Cancelled)]
+        public async Task RunTransactionAsync_RollbackNoRetryOnCommitFailure(StatusCode code)
+        {
+            var client = new TransactionTestingClient(1, CreateRpcException(code));
+            var db = FirestoreDb.Create("proj", "db", client);
+            await Assert.ThrowsAsync<RpcException>(() => db.RunTransactionAsync(CreateCountingCallback()));
+            Assert.Equal(new[] { CreateCommitRequest("transaction 1") }, client.CommitRequests);
+            Assert.Equal(new[] { CreateRollbackRequest("transaction 1") }, client.RollbackRequests);
+        }
+
+        [Theory]
+        [InlineData(StatusCode.FailedPrecondition)]
+        [InlineData(StatusCode.Internal)]
+        [InlineData(StatusCode.PermissionDenied)]
+        public async Task RunTransactionAsync_NoRollbackOnCommitFailure(StatusCode code)
+        {
+            var client = new TransactionTestingClient(1, CreateRpcException(code));
+            var db = FirestoreDb.Create("proj", "db", client);
+            await Assert.ThrowsAsync<RpcException>(() => db.RunTransactionAsync(CreateCountingCallback()));
+            Assert.Equal(new[] { CreateCommitRequest("transaction 1") }, client.CommitRequests);
+            Assert.Empty(client.RollbackRequests);
+        }
+
+        [Fact]
+        public async Task RunTransactionAsync_RollbackOnlyOnCallbackFailure()
+        {
+            var client = new TransactionTestingClient(2, CreateRpcException(StatusCode.Aborted));
+            var db = FirestoreDb.Create("proj", "db", client);
+
+            var exception = await Assert.ThrowsAsync<IOException>(() => db.RunTransactionAsync<int>(
+                async transaction =>
+                {
+                    await transaction.GetDocumentSnapshotAsync(db.Document("col/x"));
+                    throw new IOException("Bang!");
+                }));
+            Assert.Equal("Bang!", exception.Message);
+
+            Assert.Empty(client.CommitRequests);
+            Assert.Equal(new[] { CreateRollbackRequest("transaction 1") }, client.RollbackRequests);
+        }
+
+        /*[Theory]
```

Comment:
Remove?

Reply:
Whoops - reinstated instead. Changed it to a fact to make debugging work.